client/policy_validator.go

```diff
@@ -0,0 +1,348 @@
+	var names []string
+	for _, p := range p.matchingAllowed {
+		names = append(names, p.Name)
+	}
+	return names
+}
+
+func (p *PolicyValidationResult) Error() error {
+	return p.err
+}
+
+// ClusterPolicyValidator The policy validator component must be
+// used internally by the service-controller only. Client applications
+// must use the PolicyAPIClient (rest client).
+type ClusterPolicyValidator struct {
+	cli           *VanClient
+	skupperPolicy v1alpha1.SkupperClusterPolicyInterface
+	labelRegex    *regexp.Regexp
+}
+
+func NewClusterPolicyValidator(cli *VanClient) *ClusterPolicyValidator {
+	return &ClusterPolicyValidator{
+		cli:        cli,
+		labelRegex: regexp.MustCompile(ValidRfc1123Label),
+	}
+}
+
+func (p *PolicyValidationResult) addMatchingPolicy(policy v1alpha12.SkupperClusterPolicy) {
+	p.matchingAllowed = append(p.matchingAllowed, policy)
+}
+
+func (p *ClusterPolicyValidator) LoadNamespacePolicies() ([]v1alpha12.SkupperClusterPolicy, error) {
+	policies := []v1alpha12.SkupperClusterPolicy{}
+	if p.skupperPolicy == nil {
+		if p.cli.RestConfig == nil {
+			return policies, fmt.Errorf("RestConfig not defined")
+		}
+		skupperCli, err := v1alpha1.NewForConfig(p.cli.RestConfig)
+		if err != nil {
+			return policies, err
+		}
+		p.skupperPolicy = skupperCli.SkupperClusterPolicies()
+	}
+	policyList, err := p.skupperPolicy.List(v1.ListOptions{})
+	if err != nil {
+		return policies, err
+	}
+	for _, pol := range policyList.Items {
+		if p.appliesToNS(&pol) {
+			policies = append(policies, pol)
+		}
+	}
+	return policies, nil
+}
+
+func (p *ClusterPolicyValidator) AppliesToNS(policyName string) bool {
+	pol, err := p.skupperPolicy.Get(policyName, v1.GetOptions{})
+	// If policy not found, revalidate
+	if err != nil {
+		return true
+	}
+	return p.appliesToNS(pol)
+}
+
+func (p *ClusterPolicyValidator) appliesToNS(pol *v1alpha12.SkupperClusterPolicy) bool {
+	var namespaces []string
+	for _, ns := range pol.Spec.Namespaces {
+		if p.labelRegex.MatchString(ns) {
+			namespace, err := p.cli.KubeClient.CoreV1().Namespaces().Get(p.cli.Namespace, v1.GetOptions{})
+			if err == nil {
+				selector, _ := labels.Parse(ns)
+				if selector.Matches(labels.Set(namespace.Labels)) {
+					namespaces = append(namespaces, namespace.Name)
+				}
+			}
+		} else {
+			namespaces = append(namespaces, ns)
+		}
+	}
+	return utils.StringSliceContains(namespaces, "*") ||
+		utils.StringSliceContains(namespaces, p.cli.Namespace) ||
+		utils.RegexpStringSliceContains(namespaces, p.cli.Namespace)
+}
+
+func (p *ClusterPolicyValidator) Enabled() bool {
+	if p.cli.RestConfig == nil {
+		return false
+	}
+	_, err := p.LoadNamespacePolicies()
+	if err != nil && (!p.CrdDefined(err) || !p.NoPermission(err)) {
+		return false
+	}
+	return true
+}
+
+func (p *ClusterPolicyValidator) NoPermission(err error) bool {
+	if err == nil {
+		return true
+	}
+	return !strings.Contains(err.Error(), "is forbidden")
```

Comment:
I was thinking that if the error text is changed (for instance by replacing "is forbidden" with "is not allowed"), would imply also a change in the behaviour if it is not detected by the tests. What do you think about checking if the error contains some sort of alphanumeric code? I am aware that the problem would persist in the same way, but I think that error messages directed to users have more chance of being modified.

Reply:
Agreed. The code can be retrieved by the concrete error implementation.
I will update it and see if I can use the appropriate error code.